### ISSUE TITLE
TCP Listener causes memory leak when max connections are reached

### ISSUE THREAD

**Issue by rbeckman-nextgen** [nextgenhealthcare/connect#3569]
When the maximum number of connections has been reached, the next client socket that gets accepted will be added to the cached set of workers, and then the acceptor thread will try to submit it to the executor. That will fail with a RejectedExecutionException and the socket will be closed which is correct. However the worker will not be removed from the set, and will stay in memory until the connector is stopped.

Imported Issue. Original Details:
Jira Issue Key: MIRTH-3680
Reporter: narupley
Created: 2015-03-26T15:51:19.000-0700

**Comment by rbeckman-nextgen** [comment]
Revisions 7657/7658: Fixed memory leak in TCP Listener when max connections are reached. When an exception occurs while submitting the worker, it's now correctly removed from the cache.

Imported Comment. Original Details:
Author: narupley
Created: 2015-04-06T09:53:53.000-0700

**Comment by rbeckman-nextgen** [comment]
I replicated the issue in an older build by creating a TCP Listener, configuring the max incoming connections to 1. I then created 2 separate TCP Senders with a JS Reader as the source sending messages at a frequency of 100 milliseconds. I put a sys out statement within my IDE and started both channels. I was able to verify in the older build that the set of clientReaders continued to grow without bound in the older version. I then exported these channels, and in the fixed version, the readers were correctly being removed so that any one point, only the active connections were contained within the set.

Imported Comment. Original Details:
Author: jaysenp
Created: 2015-05-20T15:49:30.000-0700
